clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -210,9 +227,21 @@ void run(long now) {
         }
 
         List<ProducerBatch> expiredBatches = this.accumulator.abortExpiredBatches(this.requestTimeout, now);
-        // update sensors
-        for (ProducerBatch expiredBatch : expiredBatches)
+
+        boolean needsTransactionStateReset = false;
+        // Reset the PID if an expired batch has previously been sent to the broker. Also update the metrics
```

Comment:
Nit: I personally find comments that just repeat what the code is doing not so useful. However, if we explained why we need to reset the pid in this case, that would be pretty useful.